src/algorithms/fardetectors/FarDetectorLinearTracking.cc

```diff
@@ -0,0 +1,135 @@
+      for(auto hit : (*hits[level])){
+        auto pos = hit.getPosition();
+        hitMatrix->col(level) << pos.x, pos.y, pos.z;
+
+        if(level>0){
+          makeHitCombination(level-1,
+                             hitMatrix,
+                             hits,
+                             outputTracks);
+        }
+        else{
+          checkHitCombination(hitMatrix,outputTracks);
+        }
+      }
+
+    }
+
+
+    void FarDetectorLinearTracking::checkHitCombination(Eigen::MatrixXd* hitMatrix,
+                                                        gsl::not_null<edm4eic::TrackSegmentCollection*> outputTracks ) const {
+
+      Eigen::Vector3d weightedAnchor = (*hitMatrix)*m_layerWeights/(m_layerWeights.sum());
+
+      auto localMatrix = (*hitMatrix).colwise()-weightedAnchor;
+
+      Eigen::BDCSVD<Eigen::MatrixXd> svd(localMatrix.transpose(), Eigen::ComputeThinU | Eigen::ComputeThinV);
+
+      auto V = svd.matrixV();
+
+      // Rotate into principle components and calculate chi2/ndf
+      auto rotatedMatrix = localMatrix.transpose()*V;
+      auto residuals     = rotatedMatrix.rightCols(2);
+      double chi2        = (residuals.array()*residuals.array()).sum()/(2*m_cfg.n_layer);
+
+      if(chi2>m_cfg.chi2_max) return;
+
+      edm4hep::Vector3d outPos = weightedAnchor.data();
+      edm4hep::Vector3d outVec = V.col(0).data();
+
+      // Make sure fit was pointing in the right direction
+      if(outVec.z>0) outVec = outVec*-1;
+
+      uint64_t          surface{0};     // Surface track was propagated to (possibly multiple per detector)
+      uint32_t          system{0};      // Detector system track was propagated to
+      edm4hep::Vector3f position(outPos.x/dd4hep::mm,outPos.y/dd4hep::mm,outPos.z/dd4hep::mm);        // Position of the trajectory point [mm]
+      edm4eic::Cov3f    positionError;  // Error on the position
+      edm4hep::Vector3f momentum = {(float)outVec.x,(float)outVec.y,(float)outVec.z};       // 3-momentum at the point [GeV]
+      edm4eic::Cov3f    momentumError;  // Error on the 3-momentum
+      float             time{0};        // Time at this point [ns]
+      float             timeError{0};   // Error on the time at this point
+      float             theta = edm4eic::anglePolar(outVec);          // polar direction of the track at the surface [rad]
+      float             phi   = edm4eic::angleAzimuthal(outVec);         // azimuthal direction of the track at the surface [rad]
+      edm4eic::Cov2f    directionError;  // Error on the polar and azimuthal angles
+      float             pathlength{0};      // Pathlength from the origin to this point
+      float             pathlengthError{0}; // Error on the pathlength
+
+      edm4eic::TrackPoint point({surface,system,position,positionError,momentum,momentumError,time,timeError,theta,phi,directionError,pathlength,pathlengthError});
+      auto segment = (*outputTracks)->create(0,0);
+
+      segment.addToPoints(point);
+
+
```

Comment:
It might help with performance to not try to re-use hits that already became a track segment.

Reply:
These hits need to be kept around as sometimes a few tracks can be not quite parallel and very close together so a hit might form several of these tracks. There is no ambiguity solver in this reconstruction flow currently. What would probably help more is a very basic filter on the range of positions allowed between subsequent layers. I will look at adding this in the config.

Reply:
I've added the new configurations which should stop impossible combinations from being unnecessarily recursed into. Should reduce the combinatorics tested when full background samples are used.